Vault stores s3 secrets

daisieh opened the pull request:
Vault is now able to store AWS-style secrets for use with Minio buckets or ECS or S3 or whatever!

To test this, make sure that you have https://github.com/CanDIG/htsget_app/pull/206 running:

Write an object to your Minio store. Go to the bucket `samples` in your Minio console at https://docker.localhost:9090/buckets/samples/browse and upload https://github.com/CanDIG/htsget_app/blob/stable/data/files/NA18537.vcf.gz and https://github.com/CanDIG/htsget_app/blob/stable/data/files/NA18537.vcf.gz.tbi to the bucket.

Upload a drs object to htsget to correspond to that file, using user2's bearer token (user2 is a site admin):
```
curl -X "POST" "http://<candig_url>:5080/genomics/ga4gh/drs/v1/objects" \
     -H 'Content-Type: application/json' \
     -H 'Authorization: Bearer <user2 token>' \
     -d $'{
  "id": "NA18537.vcf.gz",
  "access_methods": [
    {
      "type": "s3",
      "access_id": "<minio_url>;samples;NA18537.vcf.gz"
    }
  ],
  "name": "NA18537.vcf.gz",
  "self_uri": "drs://localhost/NA18537.vcf.gz"
}'
```

(If you get `Vault Error`, try restarting the candigv2_vault-runner_1 container.)

Do the same for the index file: NA18537.vcf.gz.tbi.

In order for htsget to be able to read the file, though, we need to add the bucket credentials to the Vault.
Using user2's token again, get a client token for vault:
```
## vault client token
curl -X "POST" "http://<candig_url>:5080/vault/v1/auth/jwt/login" \
     -H 'Content-Type: application/json; charset=utf-8' \
     -d $'{
  "jwt": "<user2 token>",
  "role": "site_admin"
}'
```

Using that client token, get the values of minio-access-key and minio-secret-key from tmp/secrets and upload them to Vault:
```
curl -X "POST" "http://<candig_url>:5080/vault/v1/aws/<minio_url>;samples" \
     -H 'X-Vault-Token: <client token>' \
     -H 'Content-Type: application/json; charset=utf-8' \
     -d $'{
  "secret": "<minio-secret-key>",
  "access": "<minio-access-key>"
}'
```

Then you should be able to get the data back via htsget:
```
curl "http://<candig_url>:5080/genomics/htsget/v1/variants/data/NA18537" \
     -H 'Authorization: Bearer <user2 token>'
```

daisieh committed:
add a site_admin user with user attribute

daisieh committed:
Update ssl certs stuff

daisieh committed:
update minio setup ports etc

daisieh committed:
Revert "add a site_admin user with user attribute"

This reverts commit 78b52080fdbc73282e45177b2c01bc61c4ff4c54.

daisieh committed:
DIG-828: HTSget should be behind Tyk (#138)

* add htsget to tyk policy/api

* Add htsget to tyk

* add htsget stuff to .env

* Update example.env

* Revert "Update example.env"

This reverts commit ae612ec7a3e6edee8092d98c17a2405087848dd5.

daisieh committed:
minio and ssl

daisieh committed:
add some more vault auth stuff

daisieh committed:
add vault_s3_token

daisieh committed:
reorg vault so that we can do more active token refreshing

daisieh committed:
forgot opa tyk api

daisieh committed:
sleep 10 to make sure vault is up

daisieh committed:
allow vault to be accessed via tyk

daisieh committed:
interpolate files after docker image built

daisieh committed:
remove unneeded args from htsget docker-compose

daisieh committed:
pass env vars into htsget entrypoint

daisieh committed:
move creation of token to separate script

daisieh committed:
Update create_token.sh

daisieh committed:
undo bad merge changes

daisieh committed:
Update candig-data-portal

daisieh committed:
newline

daisieh committed:
crontab for vault

daisieh committed:
copy ssl files to tmp before sed

daisieh committed:
VAULT_SERVICE_URL

daisieh committed:
oops, missed one

daisieh committed:
actively wait for vault to start

daisieh committed:
if vault container has launched, wait 5 seconds more

daisieh committed:
Update entrypoint.sh

daisieh committed:
restart vault-runner

daisieh committed:
TYK_USE_SSL flag for prod

daisieh committed:
bump candig-server to version 1.6.0

daisieh committed:
katsu needs OPA_SITE_ADMIN_KEY var

shaikh-rashid commented:
You will need to rebase your branch as you are out of sync with `develop`. This may cause some conflicts in your branch as I modified a lot of the same files in https://github.com/CanDIG/CanDIGv2/pull/146

shaikh-rashid reviewed (changes requested):
Needs to be rebased

daisieh committed:
remove unneeded args

daisieh committed:
remove missed merge flags

shaikh-rashid commented on etc/env/example.env:
```diff
@@ -303,6 +304,12 @@ TYK_OPA_API_SLUG=opa
 TYK_OPA_API_LISTEN_PATH=policy
 
+## api - vault
+TYK_VAULT_API_ID=71
```
API_ID conflicts with OPA_API_ID

daisieh replied:
fixed

shaikh-rashid commented on lib/htsget-server/docker-compose.yml:
```diff
@@ -38,7 +33,13 @@ services:
     logging: *default-logging
     environment:
       HTSGET_TEST_KEY: "hoodlebug"
+      OPA_SECRET: ${CANDIG_OPA_SECRET}
```
Secrets should not be in environment variables

daisieh replied:
moved it to be a Docker secret (but I didn't test it; please make sure this works still)

daisieh replied:
I was able to test this after all...it works fine

shaikh-rashid commented on lib/vault/Dockerfile:
```diff
@@ -2,44 +2,26 @@
 # base image
 FROM alpine:3.7
```
Use alpine version from env or pin in docker-compose.yml

daisieh replied:
fixed

shaikh-rashid commented on lib/vault/Dockerfile:
```diff
@@ -2,44 +2,26 @@
+USER root
 
-# maintainer
-LABEL maintainer="brennan.brouillette@c3g.ca"
```
Update maintainer tag

daisieh replied:
done

shaikh-rashid commented on lib/vault/docker-compose.yml:
```diff
@@ -2,12 +2,7 @@ version: '3.7'
-        - VAULT_VERSION=1.9.3
-        - venv_python=${VENV_PYTHON}
-    image: ${DOCKER_REGISTRY}/vault:${VAULT_VERSION:-latest}
+    image: library/vault:latest
```
Why is there Dockerfile if not being used? Also, pin vault version unless rolling updates is acceptable.

daisieh replied:
the Dockerfile builds vault-runner.

shaikh-rashid commented on lib/vault/docker-compose.yml:
```diff
@@ -33,10 +28,40 @@ services:
       interval: 30s
       timeout: 20s
       retries: 3
+  vault-runner:
```
Why is vault runner needed?

daisieh replied:
Vault runner is needed to run the cronjob for refreshing the vault-s3-token.

daisieh replied:
That being said, the cronjob doesn't seem to work correctly. The token can be manually refreshed via the create_token.sh script, though. The cronjob can be dealt with in a future PR.

shaikh-rashid reviewed (changes requested):
Review comments

daisieh committed:
fix TYK_VAULT_API_ID

daisieh committed:
pin alpine, vault versions

daisieh committed:
maintainer

daisieh committed:
now it's just docker.localhost

Since the container is no longer specifically named auth.docker.localhost, this is just docker.localhost as well.

shaikh-rashid commented on etc/env/example.env:
```diff
@@ -326,7 +334,6 @@ VAULT_SERVICE_URL=http://vault:8200
 OPA_PORT=8181
 OPA_LOG_LEVEL=debug
-CANDIG_OPA_SECRET=my-secret-root-token
 CANDIG_OPA_SECRET_SERVICE=my-secret-service-token
```
This OPA token should also not be set from environment variable

daisieh replied:
This one is a bit trickier to fix because it needs a fix in the OPA repo as well: It doesn't really get used for anything, but I can fix it in a separate PR (pair of PRs)

shaikh-rashid commented on lib/chord-metadata/docker-compose.yml:
```diff
@@ -41,6 +41,7 @@ services:
       - CANDIG_AUTHORIZATION=${CANDIG_AUTHORIZATION}
       - CACHE_TIME=${CACHE_TIME}
       - CANDIG_OPA_SECRET=${CANDIG_OPA_SECRET}
```
Avoid using secrets in environment variables if possible

daisieh replied:
done

shaikh-rashid commented on lib/chord-metadata/docker-compose.yml:
```diff
@@ -41,6 +41,7 @@ services:
       - CANDIG_AUTHORIZATION=${CANDIG_AUTHORIZATION}
       - CACHE_TIME=${CACHE_TIME}
       - CANDIG_OPA_SECRET=${CANDIG_OPA_SECRET}
+      - CANDIG_OPA_SITE_ADMIN_KEY=${OPA_SITE_ADMIN_KEY}
```
Avoid using secrets in environment variables if possible

daisieh replied:
This one isn't a secret: it's something like "site-admin"

shaikh-rashid commented on lib/htsget-server/docker-compose.yml:
```diff
@@ -38,7 +33,14 @@ services:
     environment:
       HTSGET_TEST_KEY: "hoodlebug"
+      OPA_URL: ${OPA_URL}
+      CANDIG_AUTH: ${CANDIG_AUTHORIZATION}
```
What is the CANDIG_AUTHORIZATION field used for?

daisieh replied:
It's a flag that lets you choose whether or not the auth mechanism is OPA or something else

shaikh-rashid reviewed (changes requested):
Additional changes requested

shaikh-rashid reviewed (approved):
LGTM, great work!
